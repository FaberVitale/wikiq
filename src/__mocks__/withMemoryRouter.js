import React from "react";
import MemoryRouter from "react-router-dom/MemoryRouter";

export default (Comp) => (initialEntries = ["/"], initialIndex = 0) => {
  class MockRoute extends React.Component {
    static displayName = `withMockRoute(${
      Comp.displayName || Comp.name || "Component"
    })`;

    render() {
      return (
        <MemoryRouter
          initialEntries={initialEntries}
          initialIndex={initialIndex}
        >
          <Comp {...this.props} />
        </MemoryRouter>
      );
    }
  }
  return MockRoute;
};
