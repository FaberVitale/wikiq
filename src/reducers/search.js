//@flow
import {
  FETCH_SEARCH,
  STORE_SEARCH,
  FAILED_TO_FETCH_SEARCH,
  STORE_THUMBNAILS,
  FAILED_TO_FETCH_THUMBNAILS,
  FETCH_THUMBNAILS,
} from "../action/types";
import { PAGINATION_SIZE } from "../config";
import type { Action } from "../action/types";

export type WikiArticleInfo = {|
  +title: string,
  +description: string,
  +link: string,
|};

export type Thumbnail = {|
  +source: string,
  +width: number,
  +height: number,
|};

export type WikiArticle = {|
  +info: WikiArticleInfo,
  +thumbnail: Thumbnail | null,
|};

export type Search = {
  ids: Array<string> | null,
  fetchState: number,
  error: null | string,
  thumbnails: number,
};

export type Searches = {
  [id: string]: Search,
};

export type Articles = {
  [articleId: string]: WikiArticleInfo,
};

export type Thumbnails = {
  [articleId: string]: Thumbnail,
};

const DEFAULT_ERROR_MESSAGE = "error";

export const fetchStates = {
  IDLE: 0,
  REQUESTED_SEARCH: 1,
  REQUESTED_THUMBNAILS: 2,
};

Object.freeze(fetchStates);

const searchInitState = {
  ids: null,
  fetchState: fetchStates.IDLE,
  error: null,
  thumbnails: 0,
};

export const search = (state: Search = searchInitState, action: Action) => {
  switch (action.type) {
    case FETCH_SEARCH:
      return {
        ...state,
        fetchState: fetchStates.REQUESTED_SEARCH,
        error: null,
      };
    case STORE_SEARCH:
      return {
        ...state,
        ids: action.ids,
        fetchState: fetchStates.IDLE,
        error: null,
        thumbnails: PAGINATION_SIZE,
      };
    case FAILED_TO_FETCH_SEARCH:
      const message =
        action.error && typeof action.error.message === "string"
          ? action.error.message
          : DEFAULT_ERROR_MESSAGE;
      return {
        ...state,
        fetchState: fetchStates.IDLE,
        error: message,
      };
    case FETCH_THUMBNAILS:
      return {
        ...state,
        fetchState: fetchStates.REQUESTED_THUMBNAILS,
      };
    case STORE_THUMBNAILS:
    case FAILED_TO_FETCH_THUMBNAILS:
      return {
        ...state,
        thumbnails: state.thumbnails + PAGINATION_SIZE,
        fetchState: fetchStates.IDLE,
      };
    default:
      return state;
  }
};

export const searches = (state: Searches = {}, action: Action) => {
  switch (action.type) {
    case FAILED_TO_FETCH_SEARCH:
    case STORE_SEARCH:
    case FETCH_SEARCH:
    case STORE_THUMBNAILS:
    case FAILED_TO_FETCH_THUMBNAILS:
    case FETCH_THUMBNAILS:
      return {
        ...state,
        [action.searchId]: search(state[action.searchId], action),
      };
    default:
      return state;
  }
};

export const articles = (state: Articles = {}, action: Action) => {
  if (action.type === STORE_SEARCH) {
    const nextState = { ...state };

    const { ids, articles } = action;

    for (let i = 0, len = ids.length; i < len; i++) {
      nextState[ids[i]] = articles[i];
    }

    return nextState;
  }
  return state;
};

export const thumbnails = (state: Thumbnails = {}, action: Action) => {
  switch (action.type) {
    case STORE_SEARCH:
    case STORE_THUMBNAILS:
      return {
        ...state,
        ...action.thumbnails,
      };
    default:
      return state;
  }
};
