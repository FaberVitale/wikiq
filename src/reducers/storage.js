//@flow
import type { Reducer } from "redux";
import { $localStorage } from "../util/dom";

/* higher order reducer that retrieves the initial state from the local storage
 * If the item is present is
 * transformed using a user-defined function onValue that defaults to JSON.parse,
 *
 * It is possible to validate the result passing a function that given an input
 *
 * signature:
 *  reducer: a valid reducer,
 *  key: used to request the initial State,
 *  options:
 *    storage: an object that has getItem function,
 *      defaults to localStore if present
 *    onValue: a function that given the string value recieved applies
 *      a transformation in order to deserialize the state,
 *      defaults to JSON.parse
 *    isValid: afunction that given the rehidrated value, validates it
 *      defaults to (val: S) => true
 */
interface GetItem {
  +getItem: (key: string) => ?string;
}

type Options<S> = {
  storage: GetItem,
  isValid: (value: S) => boolean,
  onValue: (value: string) => S,
};

const fauxStorage: GetItem = {
  getItem() {
    return null;
  },
};

const defOptions: Options<any> = {
  storage: $localStorage != null ? $localStorage : fauxStorage,
  isValid: (anyVal) => true,
  onValue: JSON.parse,
};
const storageType = "@@STORAGE_INIT";

export default <S, A>(
  reducer: Reducer<S, A>,
  key: string,
  options?: $Shape<Options<S>>
) => {
  const optionsMerged: Options<S> = Object.assign({}, defOptions, options);

  let storageValue: ?string = null;
  let initState;
  let hasAcquiredInitState = false;

  try {
    storageValue = optionsMerged.storage.getItem(key);
  } catch (e) {}

  if (storageValue != null) {
    try {
      let extractedValue = optionsMerged.onValue(storageValue);

      if (optionsMerged.isValid(extractedValue)) {
        hasAcquiredInitState = true;
        initState = extractedValue;
      }
    } catch (e) {}
  }

  if (!hasAcquiredInitState) {
    initState = reducer(undefined, { action: storageType });
  }

  return (state: S = initState, action: A) => reducer(state, action);
};
