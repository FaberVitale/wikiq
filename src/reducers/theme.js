//@flow
import { CHANGE_THEME } from "../action/types";
import type { Action } from "../action/types";

export type State = "light" | "dark";

const initState: State = "light";

export default (state: State = initState, action: Action) => {
  if (action.type === CHANGE_THEME) {
    return state === "light" ? "dark" : "light";
  }
  return state;
};
