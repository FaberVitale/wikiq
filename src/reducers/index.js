//@flow
import { combineReducers } from "redux";
import theme from "./theme";
import type { State as Theme } from "./theme";
import { searches, articles, thumbnails } from "./search";
import type { Searches, Articles, Thumbnails } from "./search";

export { fetchStates } from "./search";

export type State = {|
  +theme: Theme,
  +searches: Searches,
  +articles: Articles,
  +thumbnails: Thumbnails
|};

export type { State as Theme } from "./theme";
export type { Search, Thumbnail, WikiArticle, WikiArticleInfo } from "./search";

export default combineReducers({
  theme,
  searches,
  articles,
  thumbnails
});
