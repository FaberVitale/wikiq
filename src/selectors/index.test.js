import * as selector from "./index";
import { initState, search, searchId } from "../__mocks__/reducers";

describe("src/selectors", () => {
  test("getTheme", () => {
    expect(selector.getTheme(initState)).toBe("light");
  });

  test("getSearch", () => {
    expect(selector.getSearch(initState, "")).toBeNull();
    expect(selector.getSearch(search.fetching, searchId)).toBe(
      search.fetching.searches[searchId]
    );
  });

  test("getError", () => {
    expect(selector.getError(initState, "")).toBeNull();
    expect(selector.getError(search.fetching)).toBeNull();
    expect(selector.getError(search.fetchError, searchId)).toBe(
      search.fetchError.searches[searchId].error
    );
  });

  test("hasMoreThumbnails", () => {
    expect(selector.hasMoreThumbnails(initState, "")).toBe(false);
    expect(selector.hasMoreThumbnails(search.fetched, searchId)).toBe(true);
    expect(
      selector.hasMoreThumbnails(search.fetchedAllThumbnails, searchId)
    ).toBe(false);
  });
  test("shouldFetchSearch", () => {
    expect(selector.shouldFetchSearch(initState, "")).toBe(true);
    expect(selector.shouldFetchSearch(search.fetchError)).toBe(true);
    expect(selector.shouldFetchSearch(search.fetched, searchId)).toBe(false);
    expect(
      selector.shouldFetchSearch(search.fetchedAllThumbnails, searchId)
    ).toBe(false);
    expect(selector.shouldFetchSearch(search.fetching, searchId)).toBe(false);
  });

  test("isFetchingThumbnails", () => {
    expect(selector.isFetchingThumbnails(initState, "")).toBe(false);
    expect(selector.isFetchingThumbnails(search.fetching, searchId)).toBe(
      false
    );
    expect(selector.isFetchingThumbnails(search.fetchError, searchId)).toBe(
      false
    );
    expect(
      selector.isFetchingThumbnails(search.fetchedAllThumbnails, searchId)
    ).toBe(false);
    expect(selector.isFetchingThumbnails(search.fetched, searchId)).toBe(false);
    expect(
      selector.isFetchingThumbnails(search.fetchingThumbnails, searchId)
    ).toBe(true);
  });

  test("getArticles", () => {
    expect(selector.getArticles(initState, "")).toBeNull();
    expect(selector.getArticles(initState, searchId)).toBeNull();
    expect(selector.getArticles(search.fetching, searchId)).toBeNull();
    expect(selector.getArticles(search.fetched, searchId)).toMatchSnapshot();
  });
});
