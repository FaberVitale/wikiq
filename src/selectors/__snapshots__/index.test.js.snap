// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/selectors getArticles 1`] = `
Array [
  Object {
    "info": Object {
      "description": "Rem or REM may refer to:",
      "link": "https://en.wikipedia.org/wiki/Rem",
      "title": "Rem",
    },
    "thumbnail": null,
  },
  Object {
    "info": Object {
      "description": "Rembrandt Harmenszoon van Rijn (; Dutch: [ˈrɛmbrɑnt ˈɦɑrmə(n)soːn vɑn ˈrɛin] ( listen); 15 July 1606 – 4 October 1669) was a Dutch draughtsman, painter, and printmaker.",
      "link": "https://en.wikipedia.org/wiki/Rembrandt",
      "title": "Rembrandt",
    },
    "thumbnail": Object {
      "height": 280,
      "source": "https://upload.wikimedia.org/wikipedia/commons/thumb/b/bd/Rembrandt_van_Rijn_-_Self-Portrait_-_Google_Art_Project.jpg/216px-Rembrandt_van_Rijn_-_Self-Portrait_-_Google_Art_Project.jpg",
      "width": 216,
    },
  },
  Object {
    "info": Object {
      "description": "In electronics, a remote control is a component of an electronic device used to operate the device from a distance, usually wirelessly.",
      "link": "https://en.wikipedia.org/wiki/Remote_control",
      "title": "Remote control",
    },
    "thumbnail": Object {
      "height": 202,
      "source": "https://upload.wikimedia.org/wikipedia/commons/thumb/7/78/Siri_Remote_and_old_Apple_Remote.JPG/280px-Siri_Remote_and_old_Apple_Remote.JPG",
      "width": 280,
    },
  },
  Object {
    "info": Object {
      "description": "Reminisce Mackie (née Smith; May 30, 1980), known professionally as Remy Ma, is an American rapper. She was initially discovered by Big Pun, and came to prominence for her work as a member of Fat Joe’s group, Terror Squad.",
      "link": "https://en.wikipedia.org/wiki/Remy_Ma",
      "title": "Remy Ma",
    },
    "thumbnail": Object {
      "height": 280,
      "source": "https://upload.wikimedia.org/wikipedia/commons/thumb/7/7b/Remy_MA_BET_AWARDS_2016.jpg/206px-Remy_MA_BET_AWARDS_2016.jpg",
      "width": 206,
    },
  },
  Object {
    "info": Object {
      "description": "Remembrance Day (sometimes known informally as Poppy Day) is a memorial day observed in Commonwealth of Nations member states since the end of the First World War to remember the members of their armed forces who have died in the line of duty.",
      "link": "https://en.wikipedia.org/wiki/Remembrance_Day",
      "title": "Remembrance Day",
    },
    "thumbnail": Object {
      "height": 280,
      "source": "https://upload.wikimedia.org/wikipedia/commons/thumb/1/14/Cenotaph_London.jpg/185px-Cenotaph_London.jpg",
      "width": 185,
    },
  },
  Object {
    "info": Object {
      "description": "Remington Arms Company, LLC is an American manufacturer of firearms and ammunition in the United States.",
      "link": "https://en.wikipedia.org/wiki/Remington_Arms",
      "title": "Remington Arms",
    },
    "thumbnail": null,
  },
  Object {
    "info": Object {
      "description": "A remittance is a transfer of money by a foreign worker to an individual in his or her home country. Money sent home by migrants competes with international aid as one of the largest financial inflows to developing countries.",
      "link": "https://en.wikipedia.org/wiki/Remittance",
      "title": "Remittance",
    },
    "thumbnail": Object {
      "height": 186,
      "source": "https://upload.wikimedia.org/wikipedia/commons/thumb/5/50/Remittance_advert_Oxford_Street_London_20060909.jpg/280px-Remittance_advert_Oxford_Street_London_20060909.jpg",
      "width": 280,
    },
  },
  Object {
    "info": Object {
      "description": "Remain in Light is the fourth studio album by American new wave band Talking Heads, released on October 8, 1980 through Sire Records.",
      "link": "https://en.wikipedia.org/wiki/Remain_in_Light",
      "title": "Remain in Light",
    },
    "thumbnail": null,
  },
  Object {
    "info": Object {
      "description": "The removal of cannabis from Schedule I of the Controlled Substances Act, the most tightly restricted category reserved for drugs that have \\"no currently accepted medical use,\\" has been proposed repeatedly since 1972.",
      "link": "https://en.wikipedia.org/wiki/Removal_of_cannabis_from_Schedule_I_of_the_Controlled_Substances_Act",
      "title": "Removal of cannabis from Schedule I of the Controlled Substances Act",
    },
    "thumbnail": null,
  },
  Object {
    "info": Object {
      "description": "The remembrance poppy is an artificial flower that has been used since 1921 to commemorate military personnel who have died in war, and represents a common or field poppy, Papaver rhoeas.",
      "link": "https://en.wikipedia.org/wiki/Remembrance_poppy",
      "title": "Remembrance poppy",
    },
    "thumbnail": Object {
      "height": 187,
      "source": "https://upload.wikimedia.org/wikipedia/commons/thumb/7/76/Field_poppy_-_Papaver_rhoeas_%2812190335083%29.jpg/280px-Field_poppy_-_Papaver_rhoeas_%2812190335083%29.jpg",
      "width": 280,
    },
  },
]
`;
